extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/RuntimeBeanProcessor.java

```diff
@@ -82,6 +85,30 @@ public boolean test(BeanInfo bean) {
             producer.returnValue(result);
         }
         classCreator.close();
-        recorder.initSupplierBeans(map);
+        return new BeanInfoBuildItem(staticBeans, runtimeBeans);
+    }
+
+    @BuildStep
+    @Record(ExecutionTime.STATIC_INIT)
+    void initStatic(BeanInfoBuildItem infoBuildItem,
+            ArcRecorder recorder) {
+        recorder.initStaticSupplierBeans(infoBuildItem.staticBeans);
+    }
+
+    @BuildStep
+    @Record(ExecutionTime.RUNTIME_INIT)
+    void initRuntime(BeanInfoBuildItem infoBuildItem,
+            ArcRecorder recorder) {
+        recorder.initRuntimeSupplierBeans(infoBuildItem.runtimeBeans);
+    }
+
+    static final class BeanInfoBuildItem extends SimpleBuildItem {
+        final Map<String, Supplier<Object>> staticBeans;
```

Comment:
It's minor but let's call it `staticInitBeans`. Othewise we could have the impression that these beans are static.